### PR TITLE
[6088] Scheduled publish date/time is not listed in the "Scheduled for Publish" and "Pending Approval" dashlet

### PR DESCRIPTION
**Pull request by jvega190**
- Update text for items in Scheduled for Publish and Pending Approval
- Add filter by type to dashlets

https://github.com/craftercms/craftercms/issues/6088
https://github.com/craftercms/craftercms/issues/6092

### CODE REVIEW

**Comment by jvega190** [comment]
Pending sorting capabilities in APIs for the dashlets to be properly sorted

**Review by rart — APPROVED** [review]

File: ui/app/src/components/ActionsBar/ActionsBar.tsx [hunk 1]
```diff
@@ -78,6 +80,11 @@ export function ActionsBar(props: ActionsBarProps) {
             onChange={onCheckboxChange}
           />
         )}
+        {Boolean(noSelectionContent) && (
+          <Box component="span" sx={{ display: !isLoading && !isIndeterminate && !isChecked ? 'initial' : 'none' }}>
+            {noSelectionContent}
+          </Box>
+        )}
```
Comment (rart) [inline ui/app/src/components/ActionsBar/ActionsBar.tsx]: Remove wrapper, move all conditions to the `if`

**Review by rart — CHANGES_REQUESTED** [review]
- Change "Filter by..." > "Show"
- Pluralize filter options (Items, Content Types, etc)
- DashletFilter:
  - Receives: `selectedKey: FilterSystemTypeGroups` (item, asset, contentType, etc) 
  - Emits: `onChange(e: Event, key: FilterSystemTypeGroups)`
  - `useDashletFilterState(storageKey: string): { selectedKey, onChange(), selectedTypes }`
    - Internally manages localStorage
- Services to encapsulate serialization details of the API (e.g receive array of itemTypes and service to `join(',')`)
- Backend to correct default sorting for Scheduled for Publish api  
- Remove Avatar from skeletons where they don't apply

File: ui/app/src/models/SystemType.ts [hunk 2]
```diff
@@ -24,6 +24,7 @@ export type SystemType =
   | 'taxonomy'
   | 'levelDescriptor'
   | 'file'
-  | 'unknown';
+  | 'unknown'
```
Comment (rart) [inline ui/app/src/models/SystemType.ts]: Remove unknown.

Per backend:
asset
component
content type
document
file
folder
levelDescriptor
page
renderingTemplate
script
taxonomy

File: ui/app/src/components/ActionsBar/ActionsBar.tsx [hunk 3]
```diff
@@ -78,6 +80,7 @@ export function ActionsBar(props: ActionsBarProps) {
             onChange={onCheckboxChange}
           />
         )}
+        {Boolean(noSelectionContent) && !isLoading && !isIndeterminate && !isChecked && noSelectionContent}
```
Comment (rart) [inline ui/app/src/components/ActionsBar/ActionsBar.tsx]:
```suggestion
        {!isLoading && !isIndeterminate && !isChecked && noSelectionContent}
```

File: ui/app/src/components/ActivityDashlet/DashletFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      onMenuItemClick={(e, key) => onChange(e, key as FilterSystemTypeGroups)}
+      options={options}
+      closeOnSelection={false}
+      menuProps={{ sx: { minWidth: 180 } }}
+      listItemProps={{ dense: true }}
+    >
+      {selectedKeys.includes('all') ? (
+        <FormattedMessage defaultMessage="Show" />
+      ) : (
+        <FormattedMessage defaultMessage="Filter Active ({count})" values={{ count: selectedKeys.length }} />
```
Comment (rart) [inline ui/app/src/components/ActivityDashlet/DashletFilter.tsx]:
```suggestion
        <FormattedMessage defaultMessage="Filters active: {count}" values={{ count: selectedKeys.length }} />
```

File: ui/app/src/components/ActivityDashlet/DashletFilter.tsx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      size="small"
+      variant="text"
+      onMenuItemClick={(e, key) => onChange(e, key as FilterSystemTypeGroups)}
+      options={options}
+      closeOnSelection={false}
+      menuProps={{ sx: { minWidth: 180 } }}
+      listItemProps={{ dense: true }}
+    >
+      {selectedKeys.includes('all') ? (
+        <FormattedMessage defaultMessage="Show" />
```
Comment (rart) [inline ui/app/src/components/ActivityDashlet/DashletFilter.tsx]:
```suggestion
        <FormattedMessage defaultMessage="Show all" />
```

File: ui/app/src/hooks/useDashletFilterState.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import useActiveSite from './useActiveSite';
+
+export function useDashletFilterState(storageKey: string): {
+  selectedKeys: FilterSystemTypeGroups[];
+  onChange(key: FilterSystemTypeGroups): void;
+  selectedTypes: SystemType[];
+} {
+  const { uuid } = useActiveSite();
+  const initialKeys = getDashletFilterTypeGroups(uuid, storageKey) ?? ['all'];
+  const [selectedKeys, setSelectedKeys] = useState<FilterSystemTypeGroups[]>(initialKeys);
```
Comment (rart) [inline ui/app/src/hooks/useDashletFilterState.ts]:
```suggestion
  const [selectedKeys, setSelectedKeys] = useState<FilterSystemTypeGroups[]>(() => getDashletFilterTypeGroups(uuid, storageKey) ?? ['all']);
```

File: ui/app/src/utils/state.ts [hunk 7]
```diff
@@ -324,3 +324,19 @@ export function setOutdatedXBValidationDate(siteId: string, username: string, da
   const dateString = date.toDateString();
   localStorage.setItem(`craftermcs.${username}.${siteId}.outdatedXBValidationDate`, dateString);
 }
+
+export function getDashletFilterTypeGroups(
```
Comment (rart) [inline ui/app/src/utils/state.ts]:
```suggestion
// Explain here...
export function getDashletFilterSystemTypeGroups(
```

File: ui/app/src/utils/state.ts [hunk 8]
```diff
@@ -324,3 +324,19 @@ export function setOutdatedXBValidationDate(siteId: string, username: string, da
...
+  siteIdentifier: string,
+  storageKey: string
+): (FilterSystemTypeGroups | 'all')[] {
+  const typeGroups = localStorage.getItem(`craftercms.${siteIdentifier}.${storageKey}.dashletFilterTypeGroups`);
+  return typeGroups?.split(',') as (FilterSystemTypeGroups | 'all')[];
+}
+
+export function setDashletFilterTypeGroups(
```
Comment (rart) [inline ui/app/src/utils/state.ts]:
```suggestion
// Explain..
export function setDashletFilterSystemTypeGroups(
```

File: ui/app/src/components/UnpublishedDashlet/UnpublishedDashlet.tsx [hunk 9]
```diff
@@ -224,6 +246,14 @@ export function UnpublishedDashlet(props: UnpublishedDashletProps) {
                 ]
               : [])
           ])}
+          noSelectionContent={
+            <DashletFilter
+              selectedKeys={filterState.selectedKeys}
+              onChange={(e, key) => {
+                filterState.onChange(key);
+              }}
```
Comment (rart) [inline ui/app/src/components/UnpublishedDashlet/UnpublishedDashlet.tsx]: onChange={filterState.onChange}

File: ui/app/src/components/ActivityDashlet/DashletFilter.tsx [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        selected: selectedKeys.includes(key as FilterSystemTypeGroups)
+      })),
+    [selectedKeys]
+  );
+
+  return (
+    <DropDownMenu
+      size="small"
+      variant="text"
+      onMenuItemClick={(e, key) => onChange(e, key as FilterSystemTypeGroups)}
```
Comment (rart) [inline ui/app/src/components/ActivityDashlet/DashletFilter.tsx]:
```suggestion
      onMenuItemClick={onChange}
```

File: ui/app/src/hooks/useDashletFilterState.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+import { FilterSystemTypeGroups } from '../models';
+import SystemType from '../models/SystemType';
+import { useMemo, useState } from 'react';
+import { getDashletFilterTypeGroups, setDashletFilterTypeGroups } from '../utils/state';
+import { filterOptionsLookup } from '../components/ActivityDashlet/DashletFilter';
+import useActiveSite from './useActiveSite';
+
+export function useDashletFilterState(storageKey: string): {
+  selectedKeys: FilterSystemTypeGroups[];
+  onChange(key: FilterSystemTypeGroups): void;
```
Comment (rart) [inline ui/app/src/hooks/useDashletFilterState.ts]:
```suggestion
  onChange(e: Ev, key: FilterSystemTypeGroups): void;
```

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/components/ActivityDashlet/DashletFilter.tsx [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+      onMenuItemClick={onChange}
+      options={options}
+      closeOnSelection={false}
+      menuProps={{ sx: { minWidth: 180 } }}
+      listItemProps={{ dense: true }}
+    >
+      {selectedKeys.includes('all') ? (
+        <FormattedMessage defaultMessage="Show all" />
+      ) : (
+        <FormattedMessage defaultMessage="Filters active: ({count})" values={{ count: selectedKeys.length }} />
```
Comment (rart) [inline ui/app/src/components/ActivityDashlet/DashletFilter.tsx]:
```suggestion
        <FormattedMessage defaultMessage="Filters active: {count}" values={{ count: selectedKeys.length }} />
```

File: ui/app/src/components/PendingApprovalDashlet/PendingApprovalDashlet.tsx [hunk 13]
```diff
@@ -98,47 +104,69 @@ export function PendingApprovalDashlet(props: PendingApprovalDashletProps) {
...
   const selectionOptions = useSelectionOptions(Object.values(selectedItems), formatMessage, selectedCount);
   const site = useActiveSiteId();
+  const locale = useLocale();
   const { authoringBase } = useEnv();
   const dispatch = useDispatch();
+  const filterState = useDashletFilterState('pendingApprovalDashlet');
+  const refs = useUpdateRefs({
+    currentPage,
+    filterState
+  });
```
Comment (rart) [inline ui/app/src/components/PendingApprovalDashlet/PendingApprovalDashlet.tsx]: Can we move `loadPagesUntil` here and keep a single _refs_? You could move this declaration below the others.

File: ui/app/src/components/ScheduledDashlet/ScheduledDashlet.tsx [hunk 14]
```diff
@@ -155,24 +184,34 @@ export function ScheduledDashlet(props: ScheduledDashletProps) {
...
         setState(validatedState.state);
       });
     },
-    [limit, selected, setState, site, setItemsById, loadingSkeleton]
+    [limit, selected, setState, site, setItemsById, loadingSkeleton, sortBy, sortOrder, refs]
   );
 
+  const functionRefs = useUpdateRefs({
+    loadPagesUntil
+  });
```
Comment (rart) [inline ui/app/src/components/ScheduledDashlet/ScheduledDashlet.tsx]: Same deal

File: ui/app/src/components/UnpublishedDashlet/UnpublishedDashlet.tsx [hunk 15]
```diff
@@ -96,40 +99,55 @@ export function UnpublishedDashlet(props: UnpublishedDashletProps) {
...
         setState(validatedState.state);
       });
     },
-    [limit, selected, setState, site]
+    [limit, selected, setState, site, refs]
   );
 
+  const functionRefs = useUpdateRefs({
+    loadPagesUntil
+  });
```
Comment (rart) [inline ui/app/src/components/UnpublishedDashlet/UnpublishedDashlet.tsx]: "

File: ui/app/src/utils/state.ts [hunk 16]
```diff
@@ -324,3 +324,21 @@ export function setOutdatedXBValidationDate(siteId: string, username: string, da
   const dateString = date.toDateString();
   localStorage.setItem(`craftermcs.${username}.${siteId}.outdatedXBValidationDate`, dateString);
 }
+
+// Gets the stored list of type groups used by a dashlet (identified by storageKey) for filtering
```
Comment (rart) [inline ui/app/src/utils/state.ts]:
```suggestion
// Gets the stored list of system type groups used by a dashlet (identified by storageKey) for filtering
```

File: ui/app/src/utils/state.ts [hunk 17]
```diff
@@ -324,3 +324,21 @@ export function setOutdatedXBValidationDate(siteId: string, username: string, da
...
+export function getDashletFilterSystemTypeGroups(
+  siteIdentifier: string,
+  storageKey: string
+): (FilterSystemTypeGroups | 'all')[] {
```
Comment (rart) [inline ui/app/src/utils/state.ts]: Declare this locally as a type above so you can reuse

File: ui/app/src/utils/state.ts [hunk 18]
```diff
@@ -324,3 +324,21 @@ export function setOutdatedXBValidationDate(siteId: string, username: string, da
...
+  const typeGroups = localStorage.getItem(`craftercms.${siteIdentifier}.${storageKey}.dashletFilterTypeGroups`);
+  return typeGroups?.split(',') as (FilterSystemTypeGroups | 'all')[];
```
Comment (rart) [inline ui/app/src/utils/state.ts]: Will this ever store `'all'`?

File: ui/app/src/utils/state.ts [hunk 19]
```diff
@@ -324,3 +324,21 @@ export function setOutdatedXBValidationDate(siteId: string, username: string, da
...
+}
+
+// Stores a list of group types (from where we retrieve the system types) for dashlet filtering.
```
Comment (rart) [inline ui/app/src/utils/state.ts]:
```suggestion
// Stores a list of system type groups (from where we retrieve the system types) for dashlet filtering.
```

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/utils/state.ts (hunk 17 above)
Comment (jvega190) [reply ui/app/src/utils/state.ts]: My bad, FilterSystemTypeGroups already has 'all' as it is a part of the possible values

File: ui/app/src/utils/state.ts (hunk 18 above)
Comment (jvega190) [reply ui/app/src/utils/state.ts]: yeah 'all' gets stored

**Review by rart — APPROVED** [review]

**Pull request merged**